src/libraries/System.Private.CoreLib/src/System/Number.NumberToFloatingPointBits.cs

```diff
@@ -764,5 +1551,152 @@ private static bool ShouldRoundUp(bool lsbBit, bool roundBit, bool hasTailBits)
+            // even basis, we need to round down
+            // We are only concerned with the cases where 5**q fits in single 64-bit word.
+            if ((product.low <= 1) && (q >= info.MinExponentRoundToEven) && (q <= info.MaxExponentRoundToEven) &&
+                ((mantissa & 3) == 1))
+            {
+              // we may fall between two floats!
+              // To be in-between two floats we need that in doing
+              // answer.mantissa = product.high >> (upperbit + 64 - mantissa_explicit_bits() - 3);
+              // ... we dropped out only zeroes. But if this happened, then we can go back!!!
+                if ((mantissa << (upperbit + 64 - info.DenormalMantissaBits - 3)) == product.high)
+                {
+                    // flip it so that we do not round up
+                    mantissa &= ~1UL;
+                }
+            }
+
+            mantissa += (mantissa & 1); // round up
+            mantissa >>= 1;
+            if (mantissa >= (2UL << info.DenormalMantissaBits))
+            {
+                mantissa = (1UL << info.DenormalMantissaBits);
+                // undo previous addition
+                power2++;
+            }
+
+            mantissa &= ~(1UL << info.DenormalMantissaBits);
+            if (power2 >= info.InfinitePower)
+            {
+                // infinity
+                power2 = info.InfinitePower;
+                mantissa = 0;
+            }
+            return (power2, mantissa);
+        }
+
+        private static (ulong high, ulong low) ComputeProductApproximation(int bitPrecision, long q, ulong w)
+        {
+            int index = 2 * (int)(q - SmallestPowerOfFive);
+            // For small values of q, e.g., q in [0,27], the answer is always exact because
+            // FullMultiplication gives the exact answer.
+            var firstproduct = FullMultiplication(w, s_Pow5128Table[index]);
+            ulong precision_mask = (bitPrecision < 64) ? (0xFFFFFFFFFFFFFFFFUL >> bitPrecision) : 0xFFFFFFFFFFFFFFFFUL;
+            if ((firstproduct.high & precision_mask) == precision_mask)
+            {
+              // could further guard with  (lower + w < lower)
+              // regarding the second product, we only need secondproduct.high, but our expectation is that the compiler will optimize this extra work away if needed.
+                var secondproduct = FullMultiplication(w, s_Pow5128Table[index + 1]);
+                firstproduct.low += secondproduct.high;
+                if (secondproduct.high > firstproduct.low)
+                {
+                    firstproduct.high++;
+                }
+            }
+            return firstproduct;
+        }
+
+        internal static int CalculatePower(int q)
+            => (((152170 + 65536) * q) >> 16) + 63;
+
+        internal static (ulong high, ulong low) FullMultiplication(ulong value1, ulong value2)
+        {
+            ulong hi = Math.BigMul(value1, value2, out ulong low);
```

Comment:
Is this wrapper necessary? I mean `Math.BigMul` can be used directly

Reply:
ok